odl/operator/default_ops.py

```diff
@@ -866,6 +866,321 @@ def __str__(self):
+        >>> AtAxy = op(op.adjoint(x)).inner(y)
+        >>> AtxAty = op.adjoint(x).inner(op.adjoint(y))
+        >>> AtAxy == AtxAty
+        True
+        """
+        return ComplexEmbedding(self.domain, scalar=1)
+
+
+class ImagPart(Operator):
+    def __init__(self, space):
+        """Operator that extracts imaginary part of a vector.
+
+        Parameters
+        ----------
+        space : `FnBase`
+            Space which imaginary part should be taken, needs to implement
+            ``space.real_space``.
+
+        Examples
+        --------
+        Take imaginary part of cn vector:
+
+        >>> c3 = odl.cn(3)
+        >>> op = ImagPart(c3)
+        >>> op([1 + 2j, 2, 3 - 1j])
+        rn(3).element([2.0, 0.0, -1.0])
+
+        The operator is the zero operator on real spaces
+
+        >>> r3 = odl.cn(3)
+        >>> op = ImagPart(r3)
+        >>> op([1, 2, 3])
+        rn(3).element([0.0, 0.0, 0.0])
+        """
+        real_space = space.real_space
+        linear = space == real_space
+        Operator.__init__(self, space, real_space, linear=linear)
+
+    def _call(self, x):
+        """Return ``self(x)``."""
+        return x.imag
+
+    @property
+    def inverse(self):
+        """Return the pseudoinverse.
+
+        Examples
+        --------
+        This is not a true inverse, only a pseudoinverse, the real part
+        will by necessity be lost.
+
+        >>> c3 = odl.cn(3)
+        >>> op = ImagPart(c3)
+        >>> op.inverse(op([1 + 2j, 2, 3 - 1j]))
+        cn(3).element([(2+0j), 0j, (-1+0j)])
+        """
+        return ComplexEmbedding(self.domain, scalar=1)
+
+    @property
+    def adjoint(self):
+        """Return the (left) adjoint.
+
+        Notes
+        -----
+        Due to technicalities of operators from a complex space into a real
+        space, this does not satisfy the usual adjoint equation:
+
+        .. math::
+            \langle Ax, y \rangle = \langle x, A^*y \rangle
+
+        Instead it satisfies the left adjoint equation:
+
+        .. math::
+            \langle AA^*x, y \rangle = \langle A^*x, A^*y \rangle
+
+        Examples
+        --------
+        The adjoint satisfies the adjoint equation for real spaces
+
+        >>> c3 = odl.cn(3)
+        >>> op = RealPart(c3)
+        >>> x = op.range.element([1, 2, 3])
+        >>> y = op.range.element([3, 2, 1])
+        >>> AtAxy = op(op.adjoint(x)).inner(y)
+        >>> AtxAty = op.adjoint(x).inner(op.adjoint(y))
+        >>> AtAxy == AtxAty
+        True
+        """
+        return ComplexEmbedding(self.domain, scalar=1)
+
+
+class ComplexEmbedding(Operator):
+
+    """Operator that embeds a vector into a complex space."""
+
+    def __init__(self, space, scalar=1):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        space : `FnBase`
+            Space which real part should be taken, needs to implement
+            ``space.complex_space``.
+        scalar : ``space.complex_space.field`` element
```

Comment:
I guess, but it could be `np.complex32` or `np.complex64` as well, we dont pick here.